pmlc/dialect/comp/ir/ops.td

```diff
@@ -0,0 +1,238 @@
+// Dialect
+// =============================================================================
+def COMP_Dialect : Dialect {
+  let name = "comp";
+  let cppNamespace = "pmlc::dialect::comp";
+  let extraClassDeclaration = [{
+    Type parseType(DialectAsmParser& parser) const override;
+    void printType(Type, DialectAsmPrinter& printer) const override;
+  }];
+}
+
+// =============================================================================
+// Types
+// =============================================================================
+def COMP_IsExecEnvTypePred : CPred<"$_self.isa<::pmlc::dialect::comp::ExecEnvType>()">;
+def COMP_AnyExecEnv : Type<COMP_IsExecEnvTypePred, "any execenv type">;
+
+def COMP_IsEventTypePred : CPred<"$_self.isa<::pmlc::dialect::comp::EventType>()">;
+def COMP_AnyEvent : Type<COMP_IsEventTypePred, "any event type">;
+
+class COMP_MemorySpace<string name> :
+    StrFunc<"$" # name # ".getType().cast<::mlir::MemRefType>().getMemorySpace()">;
+
+class COMP_ExecEnvSupportsMemPred<string execEnv, string memRef> :
+    CPred<"$" # execEnv # ".getType().cast<::pmlc::dialect::comp::ExecEnvType>()"
+          ".supportsMemorySpace(" # COMP_MemorySpace<memRef>.result # ")">;
+
+class COMP_ExecEnvSupportsMemTrait<string execEnv, string memRef> :
+    PredOpTrait<"memory space is supported by execenv",
+                COMP_ExecEnvSupportsMemPred<execEnv, memRef>>;
+
+// =============================================================================
+// Operations
+// =============================================================================
+class COMP_Op<string mnemonic, list<OpTrait> traits = []> :
+    Op<COMP_Dialect, mnemonic, traits>;
+
+def COMP_CreateExecEnv : COMP_Op<"create_execenv",
+                                 [MemoryEffects<[MemAlloc<DefaultResource>]>]>
+                       , Results<(outs COMP_AnyExecEnv:$execEnv)> {
+  let summary = "Creates execution environment.";
+
+  let description = [{
+    Execution environment provides functionality for scheduling operations and
+    synchronizing between them.
+  }];
+
+  let assemblyFormat = [{
+    attr-dict `:` type($execEnv)
+  }];
+}
+
+def COMP_DestroyExecEnv : COMP_Op<"destroy_execenv",
+                                  [MemoryEffects<[MemFree<DefaultResource>]>]> {
+  let summary = "Destroys execution environment.";
+
+  let description = [{
+    Any usage of environment or connected with it memory and events is invalid
+    after this operation.
+    Does not guarantee that submitted tasks finished executing.
+  }];
+
+  let arguments = (ins COMP_AnyExecEnv:$execEnv);
+
+  let assemblyFormat = [{
+     $execEnv attr-dict `:` type($execEnv)
+  }];
+}
+
+def COMP_ScheduleEnd : COMP_Op<"schedule_end", [Terminator]> {
+  let summary = "End of scheduling block.";
+}
+
+def COMP_ScheduleSingleBlock : SingleBlockImplicitTerminator<"ScheduleEnd">;
+
+def COMP_ScheduleFunc : COMP_Op<"schedule_func",
```

Comment:
Maybe "ExecFunc"?  (Scheduling is a bit overloaded in this context, and the idea of "scheduling" is already being used for ScheduleOpInterface.)

Reply:
ExecFunc to me suggests that operation executes and is finished after this operation. In general the idea is that operations named "schedule_XYZ" don't provide guarantee of finishing unless waited for, or even starting execution immediately.
Also I see no point in making a difference in naming for enqueuing/dispatching/scheduling kernel compared to any other task performed on device.